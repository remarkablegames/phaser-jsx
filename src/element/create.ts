import type { ComponentClass, FC, JSX } from 'react';

/**
 * Creates an element.
 *
 * @param type - The `type` argument must be a valid component type.
 * @param props - The `props` argument must either be an object or `null`.
 * @param children - Zero or more child elements.
 * @returns - Element object with properties `type` and `props`.
 */
export function createElement(
  type: FC | ComponentClass,
  props?: React.Attributes | null,
  ...children: JSX.Element[]
) {
  return {
    type,
    props: {
      ...props,
      children,
    },
  };
}

/**
 * Alias for `createElement`.
 *
 * {@link createElement}
 */
export const jsx = createElement;
