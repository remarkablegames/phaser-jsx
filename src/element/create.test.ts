import { createElement } from './create';

function Component() {}

it.each([
  [],
  [Component],
  [Component, null],
  [Component, null, null],
  [Component, { style: { width: 100, height: 200 } }],
  [Component, { style: { width: 100, height: 200 } }, null],
  [Component, { style: { width: 100, height: 200 } }, Component],
  [Component, {}, Component, Component],
  [Component, { children: Component }],
  [Component, { children: [Component] }],
])('creates element %p', (...args) => {
  expect(
    // eslint-disable-next-line prefer-spread
    createElement.apply(null, args as Parameters<typeof createElement>),
  ).toMatchSnapshot();
});
