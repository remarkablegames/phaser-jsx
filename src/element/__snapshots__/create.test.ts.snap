// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates element %p 1`] = `
{
  "key": null,
  "props": {},
  "type": undefined,
}
`;

exports[`creates element [Function Component] 1`] = `
{
  "key": null,
  "props": {},
  "type": [Function],
}
`;

exports[`creates element [Function Component] 2`] = `
{
  "key": null,
  "props": {},
  "type": [Function],
}
`;

exports[`creates element [Function Component] 3`] = `
{
  "key": null,
  "props": {
    "children": [
      null,
    ],
  },
  "type": [Function],
}
`;

exports[`creates element [Function Component] 4`] = `
{
  "key": null,
  "props": {
    "style": {
      "height": 200,
      "width": 100,
    },
  },
  "type": [Function],
}
`;

exports[`creates element [Function Component] 5`] = `
{
  "key": null,
  "props": {
    "children": [
      null,
    ],
    "style": {
      "height": 200,
      "width": 100,
    },
  },
  "type": [Function],
}
`;

exports[`creates element [Function Component] 6`] = `
{
  "key": null,
  "props": {
    "children": [
      [Function],
    ],
    "style": {
      "height": 200,
      "width": 100,
    },
  },
  "type": [Function],
}
`;

exports[`creates element [Function Component] 7`] = `
{
  "key": null,
  "props": {
    "children": [
      [Function],
      [Function],
    ],
  },
  "type": [Function],
}
`;

exports[`creates element [Function Component] 8`] = `
{
  "key": null,
  "props": {
    "children": [
      [Function],
    ],
  },
  "type": [Function],
}
`;

exports[`creates element [Function Component] 9`] = `
{
  "key": null,
  "props": {
    "children": [
      [Function],
    ],
  },
  "type": [Function],
}
`;
